src/client/datascience/jupyter/kernels/jupyterKernelService.ts

```diff
@@ -0,0 +1,240 @@
+import { Cancellation, wrapCancellationTokens } from '../../../common/cancellation';
+import '../../../common/extensions';
+import { traceDecorators, traceInfo } from '../../../common/logger';
+import { IFileSystem } from '../../../common/platform/types';
+
+import { ReadWrite } from '../../../common/types';
+import { noop } from '../../../common/utils/misc';
+import { IEnvironmentActivationService } from '../../../interpreter/activation/types';
+import { PythonEnvironment } from '../../../pythonEnvironments/info';
+import { captureTelemetry, sendTelemetryEvent } from '../../../telemetry';
+import { Telemetry } from '../../constants';
+import { ILocalKernelFinder } from '../../kernel-launcher/types';
+import { reportAction } from '../../progress/decorator';
+import { ReportableAction } from '../../progress/types';
+import { IJupyterKernelSpec, IKernelDependencyService } from '../../types';
+import { cleanEnvironment } from './helpers';
+import { JupyterKernelSpec } from './jupyterKernelSpec';
+import { KernelConnectionMetadata, LocalKernelConnectionMetadata } from './types';
+
+/**
+ * Responsible for registering and updating kernels
+ *
+ * @export
+ * @class JupyterKernelService
+ */
+@injectable()
+export class JupyterKernelService {
+    constructor(
+        @inject(IKernelDependencyService) private readonly kernelDependencyService: IKernelDependencyService,
+        @inject(IFileSystem) private readonly fs: IFileSystem,
+        @inject(IEnvironmentActivationService) private readonly activationHelper: IEnvironmentActivationService,
+        @inject(ILocalKernelFinder) private readonly kernelFinder: ILocalKernelFinder
+    ) {}
+
+    /**
+     * Makes sure that the kernel pointed to is a valid jupyter kernel (it registers it) and
+     * that is up to date relative to the interpreter that it might contain
+     * @param resource
+     * @param kernel
+     */
+    public async ensureKernelIsUsable(
+        kernel: KernelConnectionMetadata,
+        cancelToken?: CancellationToken,
+        disableUI?: boolean
+    ): Promise<void> {
+        // If we wish to wait for installation to complete, we must provide a cancel token.
+        const tokenSource = new CancellationTokenSource();
+        const token = wrapCancellationTokens(cancelToken, tokenSource.token);
+
+        // If we have an interpreter, make sure it has the correct dependencies installed
+        if (kernel.kind !== 'connectToLiveKernel' && kernel.interpreter) {
+            await this.kernelDependencyService.installMissingDependencies(kernel.interpreter, token, disableUI);
+        }
+
+        // If the spec file doesn't exist or is not defined, we need to register this kernel
+        if (kernel.kind !== 'connectToLiveKernel' && kernel.kernelSpec && kernel.interpreter) {
+            if (!kernel.kernelSpec.specFile || !(await this.fs.localFileExists(kernel.kernelSpec.specFile))) {
+                await this.registerKernel(kernel, token);
```

Comment:
I don't see the need for this when we're using raw & starting Python interpreters.
Else we end up registering kernels unnecessarily.

Reply:
This code is never used by anything but the jupyterSession. That's why it's named JupyterKernelService.